breadbox/breadbox/celery_task/utils.py

```diff
@@ -204,3 +204,25 @@ def update_state(
         meta["message"] = message
 
     task.update_state(state=state, meta=meta)
+
+
+def check_celery():
+    """
+    Checks to see if celery redis broker is connected.
+    Pings to see if any worker responds and returns true if successful response detected
+    """
+    inspect = app.control.inspect()
+    try:
+        # Tries to connect to celery broker
+        conn = app.broker_connection().ensure_connection(max_retries=3)
+    except Exception as exc:
+        raise CeleryConnectionError(
+            "Failed to connect to celery redis broker!"
+        ) from exc
+    # Pings workers to see if any of them respond. Returns None if no response
+    ping = inspect.ping()
```

Comment:
That's interesting... The worker is asynchronous but this call is clearly synchronous. There must be a timeout baked into this. 

Regardless if you've verified that this is reliably identifying when the worker is offline, then looks good to me.

Reply:
I looked up `inspect.ping()` and from what I see in this [SO post,](https://stackoverflow.com/questions/8506914/detect-whether-celery-is-available-running) I think any time that there's a backlog of work, will result in an exception being thrown.

Imagine we have 1 worker, and two users. User A submits a "custom analysis" job and while that's running user B submits a batch download. I think user B will get an exception saying "Celery workers are not responding" instead of A's job running and then B's job running.

Is that behavior intentional? Perhaps it's not bad, but I think at the very least, we'd want a longer timeout then the default of 1 second.

Rejecting requests when there's a backlog has some nice properties ("load shedding") but it also has the downside that when the server is busy, instead of jobs taking longer, users will have their job fail.

Reply:
I could be wrong though. It's something I think would good to be explicitly try to make sure we understand what's going to happen before deploying this change.

Reply:
@pgm yes `ping()` has a default timeout of 1 second. HOWEVER, I just saw a concerning [post](https://github.com/celery/celery/discussions/7240) where calling `ping()` may not return any response if all workers are busy running tasks. I'm thinking the `stats()` like the [top stackoverflow post](https://stackoverflow.com/questions/8506914/detect-whether-celery-is-available-running) suggested would be better

Reply:
@pgm Oops I didn't see your subsequent comments and you've basically addressed everything I was pointing out including the SO post 😅 You're right, the behavior of throwing an error when workers are occupied wasn't my intention. I think using `stats()` like the top SO reply might be what I want since it seems to return None if there are no workers present. I will test this first before making the change official